/**
 * e8yes demo web server.
 *
 * <p>Copyright (C) 2020 Chifeng Wen {daviesx66@gmail.com}
 *
 * <p>This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * <p>You should have received a copy of the GNU General Public License along with this program. If
 * not, see <http://www.gnu.org/licenses/>.
 */

syntax = "proto3";

package e8;

import "entity.proto";
import "scene_object.proto";

/*
 * @brief The SceneProto message This is a serializable version of the SceneInterface sub-classes.
 */
message SceneProto {
    /**
     * @brief The StructureType enum Possible data structures for organizing the scene content.
     */
    enum StructureType {
        INVALID = 0;

        // A linear (flat) scene structure.
        LINEAR = 1;

        // An octree scene structure.
        OCTREE = 2;
    }

    // Id of the scene.
    string id = 1;

    // A descriptive human readable name of the scene.
    string name = 2;

    // To distinguish which scene sub-class should be used for the content.
    StructureType structure_type = 3;

    // The backgr0und color of the scene. The array has exactly 3 elements.
    repeated float background_color = 4;

    // A collection of all scene objects.
    repeated SceneObject objects = 5;

    // A collection of scene entities.
    SceneEntityCollection entities = 6;
}
